PR: Add more content to the debugger and breakpoint docs and update them for Spyder 4 (#208)
spyder-ide/spyder-docs: doc/debugging.rst

Debugging with ipdb
===================

You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

> .. image:: images/debugging/debugging-commands.gif

Reviewer: This otherwise looks good, but after stepping the GIF loops back to the beginning almost instantly, giving the user no time to actually take in what just happened and leaving them confused that things jumped. Instead, I'd pause for probably a full two seconds to allow them to see and understand the effect of the next command before it jumps back.

Reviewer 2: Agreed.

Reviewer: This is now fixed, thanks.

   :alt: A Spyder IPython console window, showing debugging from toolbar

Additionally, the  :doc:`editor` shows the line of code the debugger is currently stopped on with an arrow.

.. _console commands: https://pythonconquerstheuniverse.wordpress.com/2009/09/10/debugging-in-python/

.. image:: images/debugging/debugging-arrow.png
   :alt: Spyder Editor showing the debugging panel

> 
> 

Reviewer: ```suggestion

```

Spyder's debugger offers syntax highlighting, code completion and command history, which work exactly like they do in the normal interactive interpreter.

Use the up and down arrows to recall previous commands, and press :kbd:`Tab` to trigger autocomplete suggestions.

> .. image:: images/debugging/debugging-autocompletion.gif

Reviewer: Not sure you need to show typing the `import numpy as np` line as part of the animated GIF unless typing it illustrates something I'm missing, just takes up extra time and space. Instead, start with it already entered, like you do with `li` in the next GIF.

Reviewer 2: This one is fine for me, so it's up to you @juanis2112 if you want to follow @CAM-Gerlach's suggestion or not.

Reviewer: This is addressed, thanks.

   :alt: A Spyder IPython console window, showing autocompletion when debugging

> 
> 

Reviewer: ```suggestion

```

Furthermore, IPython's `magic functions`_ are available in debugging mode.
You can, for example, run ``%ls`` to list the contents of your current working directory or ``%timeit`` to check how fast a given snippet of code is.

.. _magic functions: https://ipython.readthedocs.io/en/stable/interactive/magics.html

.. image:: images/debugging/debugging-timeit.png
   :alt: Ipython console in debug mode showing timeit magic

Finally, you can enter and execute multiline statements in Spyder's debugger just like with the regular IPython prompt, to easily run complex code.

> .. image:: images/debugging/debugging-multiline.gif

Reviewer: There's a bug in the code typed here, which the GIF makes clear: it prints `li` instead of `number`, printing the whole list three times instead of each number that you're iterating over, defeating the point of the loop. Instead, it should be `print(number)`.

Reviewer 2: Agreed.

Reviewer: This is now fixed, thanks.

   :alt: Debugger showing multiline debugger

===========
Breakpoints
===========

Spyder's debugger is integrated with the :guilabel:`Breakpoints` pane, which lists the file, line, and condition (if any) of every breakpoint defined.
To open it, select :menuselection:`Debug --> List breakpoints`, or press :kbd:`Ctrl-Shift-B` (:kbd:`Cmd-Shift-B` on macOS).

.. image:: images/debugging/breakpoints-standard.png
   :alt: Spyder's Breakpoints panel, with a number of examples showing file, line number and an optional condition

There are several different ways to set and clear breakpoints:

  * With the :guilabel:`Set/clear breakpoint` option in the Debug menu.
  * Through pressing the configurable keyboard shortcut (:kbd:`F12` for normal, or :kbd:`Shift-F12` for conditional breakpoints by default).
  * By clicking to the left of the line number in an open file in the Editor (adding :kbd:`Shift` for a conditional breakpoint).
  * With the ``breakpoint()`` builtin function in your code.
  * Interactively, using the ``b`` command in a debugging session.

> .. image:: images/debugging/debugging-breakpoints.gif

Reviewer: * [ ] It would be clearer, more useful, and more typical, to show a conditional breakpoint being set inside one of the for loops you have there, to break execution at a certain loop iteration. So you could do, e.g., `b == 4`; assuming you're calling `millerRabin` with argument e.g. `m >= 6`. 
* [ ] I also think it useful to actually run the debugger with the Variable Explorer open) to show that it is stopped when `b` equals that value; it should only take a few more seconds to run (just press the debug button and open the Variable Explorer, showing the `b` value) and would illustrate the effect of the condition breakpoint.

Reviewer 2: Agreed with both suggestions.

Reviewer: This looks good to me now, thanks.

   :alt: Spyder showing setting conditional breakpoint

> 
> 

Reviewer: ```suggestion

```

You can access and edit local and global variables at each breakpoint through the :doc:`variableexplorer`.

.. image:: images/debugging/debugging-variables.png
   :alt: Spyder's console and variable explorer showing local and global variables when debugging

> 
> 

Reviewer: ```suggestion

```

=================
Advanced features
=================

You can avoid stepping through other Python packages while debugging by enabling the new :guilabel:`Ignore Python libraries while debugging` option in Spyder's preferences, under :menuselection:`IPython Console --> Debugger --> Debug`.
This will skip all the built-in and third-party Python modules you have installed.

.. image:: images/debugging/debugging-libraries.png
   :alt: Spyder's preferences showing Ignore Python libraries while debugging options

If your code has variables with the same names as Pdb commands (e.g. ``b`` or ``step``), you can still refer to them as normal while debugging.
To call the respective Pdb command, just add an exclamation point before it (e.g. ``!b`` or ``!step``).

.. image:: images/debugging/debugging-commands.png

>    :alt: Spyder's ipython console showing Pdb commands
> 
> 

Reviewer: ```suggestion
   :alt: Spyder's IPython console showing Pdb commands

```

You can have Spyder automatically execute a custom snippet of code every time the debugger stops.
For example, you can use this to set specific variables, or import commonly-used modules so they are always available while debugging.
To set this up, go to :menuselection:`Preferences --> IPython Console --> Debugger --> Run code while debugging`, and enter the code that you want to be executed with each step.

.. image:: images/debugging/debugging-snippet.png
   :alt: Spyder's preferenes with debugging snippet

> 
> 

Reviewer: ```suggestion

```

GIF: Browsing autocomplete suggestions

You can also take advantage of IPython's `magic functions`_ in debugging mode which will allow you to use ``%matplotlib`` to switch between the ``qt5`` and ``inline`` plotting backends, and ``%timeit`` to check how fast a given snippet of code is.

.. _magic functions: https://ipython.readthedocs.io/en/stable/interactive/magics.html

SCREENSHOT: (timeit)

Additionally the debugger allows highlighting of the current frame (debugging step) in the :doc:`editor`.

SCREENSHOT of highlighting step (close up)

Finally, you can write multiline statements while debugging which will allow you execute code while debugging easily.

GIF of multiline debugging

===========
Breakpoints
===========

Spyder's debugger is integrated with a :guilabel:`Breakpoints` pane, listing the file, line, and condition (if any) of every breakpoint defined (:menuselection:`Debug --> List breakpoints`, or :kbd:`Ctrl-Shift-B` by default).

.. image:: images/debugging/breakpoints-standard.png
   :alt: Spyder's Breakpoints panel, with a number of examples showing file, line number and an optional condition

> There are multiple means of setting and clearing normal and conditional breakpoints for any line in a file opened in the :doc:`editor`.

Reviewer: ```suggestion
There are several different ways to set and clear breakpoints:
```

Greatly simplify my original overly complex, run-on sentence and make it more action- and reader-focused, and remove extraneous material ("any line in the editor")

>   * By selecting the respective option from the Debug menu.

Reviewer: ```suggestion
  * With the :guilabel:`Set/clear breakpoint` option in the Debug menu.
```

Fix my writing to be more specific and less obtuse here

>   * Through pressing a configurable keyboard shortcut (:kbd:`F12` for normal, or :kbd:`Shift-F12` for conditional breakpoints by default).

Reviewer: ```suggestion
  * Through pressing the configurable keyboard shortcut (:kbd:`F12` for normal, or :kbd:`Shift-F12` for conditional breakpoints by default).
```

Wording tweak

>   * By clicking to the left of the line number in an open file in the Editor and :kbd:`Shift`-clicking for conditional breakpoints.

Reviewer: ```suggestion
  * By clicking to the left of the line number in an open file in the Editor (adding :kbd:`Shift` for a conditional breakpoint).
```

Avoid repetition, grammar inconsistencies and awkward phrasing

>   * With the ``breakpoint()`` function in your code.

Reviewer: ```suggestion
  * With the ``breakpoint()`` builtin function in your code.
```

Clarify this is a builtin, rather than something you need to import or write yourself

Reviewer: FYI @juanis2112, not sure if you were waiting on @ccordoba12 on this one or it was just an oversight

  * Interactively, using the ``!b`` command in an ``ipdb`` session.

GIF: setting conditional breakpoint shift-clicking

You can access and edit local and global variables at each breakpoint through the :doc:`variableexplorer`.

SCREENSHOT with console and variable explorer showing local and global variables

=================
Advanced features
=================

> You can avoid stepping through other Python packages while debugging by enabling the new :guilabel:`Ignore Python libraries while debugging` option in Spyder's preferences under :menuselection:`IPython Console --> Debugger --> Debug`.

Reviewer: ```suggestion
You can avoid stepping through other Python packages while debugging by enabling the new :guilabel:`Ignore Python libraries while debugging` option in Spyder's preferences, under :menuselection:`IPython Console --> Debugger --> Debug`.
```

Needs a comma to break up the runon

This will skip all the built-in and third-party Python modules.

SCREENSHOT with option in preferences

> When debugging, if your code has variables with the name of Pdb commands, your debugging workflow might be interrupted.
> To avoid the mix of Pdb commands and Python statements when using the interactive prompt add an exclamation mark in front of the Pdb commands to separate them from Python expressions.

Reviewer: ```suggestion
If your code has variables with the same names as Pdb commands (e.g. ``a``, ``step``), you can still refer to them as normal while debugging.
To call the Pdb command, just add an exclamation point (e.g. ``!a``, ``!step``).
```

* Greatly simplify and clarify this part by cutting out a large amount of repetition and redundancy
* Be more specific and descriptive
* Fix grammar and syntax issues
* Add concrete examples

SCREENSHOT with "b" variable

> You can execute a snippet of code before any Pdb command to make sure you have the modules and variables in your context when debugging.

Reviewer: ```suggestion
You can have Spyder automatically execute a custom snippet of code every time the debugger stops.
For example, you can use this to set specific variables, or import commonly-used modules so they are always available while debugging.
```

* Be more specific about what is executed, when it is executed, what is doing the executing
* Break the example into a separate line to avoid a run-on sentence, clarify that it is an example and be more concrete about how it can be used
* Clarify that it is not run every command, just every time the prompt returns (as far as I understand it)

> To set this up, go to :menuselection:`Preferences --> IPython Console --> Debugger --> Run code while debugging` and write there the code that you want to be executed before the Pdb commands.

Reviewer: ```suggestion
To set this up, go to :menuselection:`Preferences --> IPython Console --> Debugger --> Run code while debugging`, and enter the code that you want to be executed with each step.
```

* Clarify when the code gets executed, as above (suggestions welcome if you have a better way to phrase this)
* Fix some phrasing issues
* Add a needed comma

SCREENSHOT with Run code while debugging in preferences.

==================
Matplotlib support
==================

> Plotting Matplotlib figures is fully supported while debugging with the :guilabel:`inline` backend and the :guilabel:`interactive` backends which you can use by typing the ``%matplotlib`` magic.

Reviewer: ```suggestion
Generating Matplotlib figures is fully supported while the debugger is active, including all the different graphics backends.
Use the ``%matplotlib`` magic to change to an interactive backend (e.g. ``%matplotlib qt5``) to pan, zoom and adjust your plots in a separate window, or switch back to the default ``inline`` (``%matplotlib inline``) to see them displayed right in the console.
```

* Greatly improve overall phrasing
* Be much more specific about the commands to use and why you would use each backend
* Break up this run-on sentence into more digestible parts
* Avoid much repetition/redundancy and convoluted synatax

Debugging with ipdb
===================

> You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

Reviewer: We don't do this for the other panes with shortcuts; they are all listed in the shortcut summary GUI and in preferences, along with on the tooltips of their toolbar buttons and next to their respective commands in the Debug menu. Instead of a bespoke table, if we do want to show something it would be better to just show a screenshot of the Debug menu, since it shows not only the commands and the shortcuts, but also their icons and where they can be found in the interface.

> You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

Reviewer 2: Ok, sounds good to me. @juanis2112, what do you think?

> You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

Reviewer: I don't see this added yet

> You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

Author: I realised there is a gif right under this sentence. Not sure if the screenshot fits somewhere else. What do you think?

> You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

Reviewer: Good point, I don't see anywhere it could easily fit. The one thing you could do would be to move the GIF above this paragraph (which would make it at the very top of the section) and the screenshot below, but that's not ideal and would lead to a high screenshot density. My suggestion would just be to not add it, since its not critical (since users can always just click the debug menu themselves to see it, if they need to), but I'm not sure what @ccordoba12 thinks.

> You can fully control debugger execution from the :guilabel:`Debug` menu, :guilabel:`Debug toolbar` and via configurable keyboard shortcuts, along with the standard ``ipdb`` `console commands`_.

Reviewer 2: > My suggestion would just be to not add it

Ok, I'm fine with that, so let's leave it out.

Additionally, the  :doc:`editor` shows the line of code the debugger is currently stopped on with an arrow.

.. _console commands: https://pythonconquerstheuniverse.wordpress.com/2009/09/10/debugging-in-python/

> GIF: Inset of Spyder's Editor, with a breakpoint set and the condition dialog open (commands from the toolbar)

Reviewer 2: I think we agreed that at this point we are going to show a gif with

* A breakpoint set in the editor
* Start the debugger from the toolbar
* Press the `next line` button once to show the blue arrow on the next line after the breakpoint.

Reviewer: I don't recall the specifics on this, so I'm not sure...

> Spyder's debugger offers syntax highlighting, code completion and command history, which work exactly like they do in the normal interactive interpreter.

Reviewer: This was how it read originally (minus the typo), but I proposed a the change to the current version which Juanita accepted, because that phrasing really isn't accurate. Debugging still occurs in the IPython console pane, so it doesn't make sense to say that those features work the same they do in the IPython console they still _are_ in the IPython console, and saying that implies that the debugger is in a separate pane from the console, which is potentially confusing and misleading to users.

That said, I'm not entirely happy with "normal interactive interpreter" in terms of being easily understandable, so (as I mentioned in my original comment) I'd encourage your suggestions if you have a clearer term to use here.

> Spyder's debugger offers syntax highlighting, code completion and command history, which work exactly like they do in the normal interactive interpreter.

Reviewer 2: Ok, you're right. I don't have other proposals, so let's leave it like this.

> Spyder's debugger offers syntax highlighting, code completion and command history, which work exactly like they do in the normal interactive interpreter.

Reviewer: Okay, sounds good

Use the up and down arrows to recall previous commands, and press :kbd:`Tab` to trigger autocomplete suggestions.

GIF: Browsing autocomplete suggestions

Furthermore, IPython's `magic functions`_ are available in debugging mode.

> You can, for example, run ``%matplotlib <BACKEND>`` to switch between the ``qt5`` and ``inline`` plotting backends, or ``%timeit`` to check how fast a given snippet of code is.

Reviewer: Yeah, that bothered me too, I just didn't get around to replacing it with something else.

.. _magic functions: https://ipython.readthedocs.io/en/stable/interactive/magics.html

SCREENSHOT: (timeit)

> SCREENSHOT of highlighting step (close up)

Reviewer 2: I don't understand why we need to show that screenshot at this point (it's a bit confusing to me).

Reviewer: Its a simple editing oversight; this should have gotten moved above along with the text that refers to it. Also, to note its an icon, not "highlighting" (that's in Spyder 3).

Author: Yeap. Sorry about that

> Finally, you can enter and execute multiline statements in Spyder's debugger just like with the regular IPython prompt, to easily execute complex code while debugging.

Reviewer 2: ```suggestion
Finally, you can enter and execute multiline statements in Spyder's debugger just like with the regular IPython console, to easily execute complex code.
```

* Repeating again `IPython console` for familiarity
* Removing `while debugging` because you mentioned before that those commands are entered in `Spyder's debugger`.

Reviewer: > Repeating again IPython console for familiarity

I changed this for the same reason I mentioned above. They are still _in_ the IPython console (the Spyder pane, as the both the UI and these docs consistently use to refer to it), just not using the normal IPython REPL, so it doesn't make sense to say just "like" it. However, if you have a better word than "prompt" (or "interpreter", REPL, etc) I'm all ears.

For what its worth, this would have been another good reason to simplify the name to just the `Console`, but we already fought that battle...

> Removing while debugging because you mentioned before that those commands are entered in Spyder's debugger.

Yeah, good point; I also noticed a repetition of `execute`. So, we thus have

```suggestion
Finally, you can enter and execute multiline statements in Spyder's debugger just like with the regular IPython prompt, to easily run complex code.
```

GIF of multiline debugging

===========
Breakpoints
===========

Spyder's debugger is integrated with the :guilabel:`Breakpoints` pane, which lists the file, line, and condition (if any) of every breakpoint defined.

> To open it, select :menuselection:`Debug --> List breakpoints`, or press :kbd:`Ctrl-Shift-B` by default.

Reviewer 2: ```suggestion
To open it, select :menuselection:`Debug --> List breakpoints`, or press :kbd:`Ctrl-Shift-B` (:kbd:`Cmd-Shift-B` in macOS).
```

* `by default` sounds a bit odd.
* Add shortcut for Mac.

Reviewer: I agree it sounds a little weird in that context, even if I do understand why its there.

Good catch on macOS, but in this context, its "on" macOS, not "in" (it can be tricky to know which in these cases).

```suggestion
To open it, select :menuselection:`Debug --> List breakpoints`, or press :kbd:`Ctrl-Shift-B` (:kbd:`Cmd-Shift-B` on macOS).
```

.. image:: images/debugging/breakpoints-standard.png
   :alt: Spyder's Breakpoints panel, with a number of examples showing file, line number and an optional condition

There are several different ways to set and clear breakpoints:

  * With the :guilabel:`Set/clear breakpoint` option in the Debug menu.
  * Through pressing the configurable keyboard shortcut (:kbd:`F12` for normal, or :kbd:`Shift-F12` for conditional breakpoints by default).
  * By clicking to the left of the line number in an open file in the Editor (adding :kbd:`Shift` for a conditional breakpoint).
  * With the ``breakpoint()`` function in your code.

>   * Interactively, using the ``!b`` command in an ``ipdb`` session.

Reviewer 2: ```suggestion
  * Interactively, using the ``b`` command in an debugging session.
```

* This comes before the explanation about using `!` for Pdb commands, so I think we should mention the command all people know.
* `ipdb session` sounds too technical to me.

Reviewer: > This comes before the explanation about using ! for Pdb commands, so I think we should mention the command all people know.

I thought we already talked about this and decided on ``!b``, since if users have a variable with the same name (not unlikely, especially for beginners who are unlikely to know the ipdb commands anyway or know what to do when something goes wrong), it won't work. And if "all users" already know the command (which I rather doubt), they don't need to be told it.

Also, fixed grammar error resulting from the ("an" -> "an"

```suggestion
  * Interactively, using the ``!b`` command in a debugging session.
```

Reviewer 2: I really disagree with using `!b` here. It'd be confusing for people that know Pdb commands (it'd look to them as a typo).

If people type `b` and don't see their breakpoints, they'll get the informative message we added telling them what to do. So I think it's better to leave the usage of `!` to its own section below.

Reviewer: Okay, I guess that makes sense. Therefore, with that change, the grammar-corrected revision,

```suggestion
  * Interactively, using the ``b`` command in a debugging session.
```

GIF: setting conditional breakpoint shift-clicking

You can access and edit local and global variables at each breakpoint through the :doc:`variableexplorer`.

SCREENSHOT with console and variable explorer showing local and global variables

=================
Advanced features
=================

You can avoid stepping through other Python packages while debugging by enabling the new :guilabel:`Ignore Python libraries while debugging` option in Spyder's preferences, under :menuselection:`IPython Console --> Debugger --> Debug`.

> This will skip all the built-in and third-party Python modules.

Reviewer 2: ```suggestion
This will skip all the built-in and third-party Python modules you have installed.
```

* Small clarification

SCREENSHOT with option in preferences

> If your code has variables with the same names as Pdb commands (e.g. ``a``, ``step``), you can still refer to them as normal while debugging.

Reviewer 2: ```suggestion
If your code has variables with the same names as Pdb commands (e.g. ``b`` or ``step``), you can still refer to them as normal while debugging.
```

* We agreed on using `b` for the screenshot below.
* `or` is missing to me.

Reviewer: The or isn't necessary in this context (ordered list of examples), but its not wrong to have it with it

> To call the Pdb command, just add an exclamation point (e.g. ``!a``, ``!step``).

Reviewer 2: ```suggestion
To call the respective Pdb command, just add an exclamation point before it (e.g. ``!b`` or ``!step``).
```

* I think `respective` is missing to make the connection with the previous sentence.
* I also added `before it` to be really explicit about what needs to be done (because this is a unique feature to Spyder).

SCREENSHOT with "b" variable

You can have Spyder automatically execute a custom snippet of code every time the debugger stops.
For example, you can use this to set specific variables, or import commonly-used modules so they are always available while debugging.
To set this up, go to :menuselection:`Preferences --> IPython Console --> Debugger --> Run code while debugging`, and enter the code that you want to be executed with each step.

> SCREENSHOT with Run code while debugging in preferences.

Reviewer 2: I suggest setting `import matplotlib.pyplot as plt` in `Lines` for this screenshot. That would make a nice connection with the next section.

==================
Matplotlib support
==================

Generating Matplotlib figures is fully supported while the debugger is active, including all the different graphics backends.

> Use the ``%matplotlib`` magic to change to an interactive backend (e.g. ``%matplotlib qt5``) to pan, zoom and adjust your plots in a separate window, or switch back to the default ``inline`` (``%matplotlib inline``) to see them displayed right in the console.

Reviewer 2: ```suggestion
Use the ``%matplotlib`` magic to change to an interactive backend (e.g. ``%matplotlib qt5``) to pan, zoom and adjust your plots in a separate window, or switch back to the default ``inline`` (``%matplotlib inline``) to see them displayed right in the :doc:`plot` pane.
```

* That's the default behavior.

Reviewer: Ah, didn't know if that still worked while debugging (the text before led me to believe otherwise).
